File: threads/src/main/java/im/threads/internal/utils/ConsultWriter.java
package im.threads.internal.utils;

import android.content.SharedPreferences;

import im.threads.business.models.ConsultConnectionMessage;
import im.threads.business.models.ConsultInfo;

public final class ConsultWriter {
    private static final String OPERATOR_STATUS = "OPERATOR_STATUS";
    private static final String OPERATOR_NAME = "OPERATOR_NAME";
    private static final String OPERATOR_TITLE = "OPERATOR_TITLE";
    private static final String OPERATOR_ORG_UNIT = "OPERATOR_ORG_UNIT";
    private static final String OPERATOR_ROLE = "OPERATOR_ROLE";
    private static final String OPERATOR_PHOTO = "OPERATOR_PHOTO";
    private static final String OPERATOR_ID = "OPERATOR_ID";
    private static final String SEARCHING_CONSULT = "SEARCHING_CONSULT";
    private final SharedPreferences sharedPreferences;

    /**
     * @param sharedPreferences only secured shared preferences are allowed!
     */
    public ConsultWriter(SharedPreferences sharedPreferences) {
        this.sharedPreferences = sharedPreferences;
    }

    public void setSearchingConsult(boolean isSearching) {
        sharedPreferences.edit().putBoolean(ConsultWriter.class + SEARCHING_CONSULT, isSearching).commit();
    }

    public boolean isSearchingConsult() {
        return sharedPreferences.getBoolean(ConsultWriter.class + SEARCHING_CONSULT, false);
    }

    public void setCurrentConsultInfo(ConsultConnectionMessage message) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        String consultId = message.getConsultId();
        editor
                .putString(OPERATOR_ID, consultId)
                .putString(OPERATOR_STATUS + consultId, message.getStatus())
                .putString(OPERATOR_NAME + consultId, message.getName())
                .putString(OPERATOR_TITLE + consultId, message.getTitle())
                .putString(OPERATOR_ORG_UNIT + consultId, message.getOrgUnit())
                .putString(OPERATOR_ROLE + consultId, message.getRole())
                .putString(OPERATOR_PHOTO + consultId, message.getAvatarPath())
                .commit();
    }

    public String getName(String id) {
        return sharedPreferences.getString(OPERATOR_NAME + id, null);
    }

    private String getStatus(String id) {
        return sharedPreferences.getString(OPERATOR_STATUS + id, null);
    }

    private String getOrgUnit(String id) {
        return sharedPreferences.getString(OPERATOR_ORG_UNIT + id, null);
    }

    private String getRole(String id) {
        return sharedPreferences.getString(OPERATOR_ROLE + id, null);
    }

    private String getPhotoUrl(String id) {
        return sharedPreferences.getString(OPERATOR_PHOTO + id, null);
    }

    public String getCurrentPhotoUrl() {
        return getCurrentConsultId() != null ? getPhotoUrl(getCurrentConsultId()) : null;
    }

    public String getCurrentConsultId() {
        return sharedPreferences.getString(OPERATOR_ID, null);
    }

    public void setCurrentConsultLeft() {
        sharedPreferences.edit().putString(OPERATOR_ID, null).commit();
    }

    public boolean isConsultConnected() {
        return sharedPreferences.getString(OPERATOR_ID, null) != null;
    }

    public ConsultInfo getConsultInfo(String id) {
        return new ConsultInfo(getName(id), id,
                getStatus(id), getOrgUnit(id), getRole(id), getPhotoUrl(id));
    }

    public ConsultInfo getCurrentConsultInfo() {
        String currentId = getCurrentConsultId();
        if (currentId == null) {
            return null;
        } else {
            return getConsultInfo(currentId);
        }
    }
}

File: src/main/java/im/threads/business/utils/ConsultWriter.kt
package im.threads.business.utils

import im.threads.business.models.ConsultConnectionMessage
import im.threads.business.models.ConsultInfo
import im.threads.business.preferences.Preferences

class ConsultWriter(private val preferences: Preferences) {
    var isSearchingConsult: Boolean
        get() = preferences.get("${ConsultWriter::class.java}$SEARCHING_CONSULT}") ?: false
        set(isSearching) {
            preferences.save("${ConsultWriter::class.java}$SEARCHING_CONSULT}", isSearching)
        }

    fun setCurrentConsultInfo(message: ConsultConnectionMessage) {
        val consultId = message.consultId
        preferences.save(OPERATOR_ID, consultId)
        preferences.save(OPERATOR_STATUS + consultId, message.status)
        preferences.save(OPERATOR_NAME + consultId, message.name)
        preferences.save(OPERATOR_TITLE + consultId, message.title)
        preferences.save(OPERATOR_ORG_UNIT + consultId, message.orgUnit)
        preferences.save(OPERATOR_ROLE + consultId, message.role)
        preferences.save(OPERATOR_PHOTO + consultId, message.avatarPath)
    }

    private fun getName(id: String): String? {
        return preferences.get("$OPERATOR_NAME$id")
    }

    private fun getStatus(id: String): String? {
        return preferences.get("$OPERATOR_STATUS$id")
    }

    private fun getOrgUnit(id: String): String? {
        return preferences.get("$OPERATOR_ORG_UNIT$id")
    }

    private fun getRole(id: String): String? {
        return preferences.get("$OPERATOR_ROLE$id")
    }

    private fun getPhotoUrl(id: String?): String? {
        return preferences.get("$OPERATOR_PHOTO$id")
    }

    fun getCurrentPhotoUrl(): String? {
        return if (getCurrentConsultId() != null) {
            getPhotoUrl(getCurrentConsultId())
        } else {
            null
        }
    }

    fun getCurrentConsultId(): String? {
        return preferences.get(OPERATOR_ID)
    }

    fun setCurrentConsultLeft() {
        preferences.save<String>(OPERATOR_ID, null)
    }

    fun isConsultConnected(): Boolean {
        return !preferences.get<String>(OPERATOR_ID).isNullOrBlank()
    }

    fun getConsultInfo(id: String): ConsultInfo {
        return ConsultInfo(
            getName(id),
            id,
            getStatus(id),
            getOrgUnit(id),
            getRole(id),
            getPhotoUrl(id)
        )
    }

    fun getCurrentConsultInfo(): ConsultInfo? {
        val currentId = getCurrentConsultId()
        return currentId?.let { getConsultInfo(it) }
    }

    companion object {
        private const val OPERATOR_STATUS = "OPERATOR_STATUS"
        private const val OPERATOR_NAME = "OPERATOR_NAME"
        private const val OPERATOR_TITLE = "OPERATOR_TITLE"
        private const val OPERATOR_ORG_UNIT = "OPERATOR_ORG_UNIT"
        private const val OPERATOR_ROLE = "OPERATOR_ROLE"
        private const val OPERATOR_PHOTO = "OPERATOR_PHOTO"
        private const val OPERATOR_ID = "OPERATOR_ID"
        private const val SEARCHING_CONSULT = "SEARCHING_CONSULT"
    }
}
